src/pybind/mgr/prometheus/module.py

```diff
@@ -1068,34 +1117,66 @@ def index(self):
-                    instance.collect_lock.release()
+                assert isinstance(_global_instance, Module)
+                with _global_instance.collect_lock:
+                    return self._metrics(_global_instance)
 
             @staticmethod
             def _metrics(instance):
-                # Return cached data if available and collected before the
-                # cache times out
-                if instance.collect_cache and time.time() - instance.collect_time < instance.collect_timeout:
-                    cherrypy.response.headers['Content-Type'] = 'text/plain'
-                    return instance.collect_cache
+                # type: (Module) -> Any
+                # Return cached data if available
+                if not instance.collect_cache:
+                    raise cherrypy.HTTPError(503, 'No cached data available yet')
 
-                if instance.have_mon_connection():
-                    instance.collect_cache = None
-                    instance.collect_time = time.time()
-                    instance.collect_cache = instance.collect()
+                def respond():
+                    assert isinstance(instance, Module)
                     cherrypy.response.headers['Content-Type'] = 'text/plain'
                     return instance.collect_cache
-                else:
-                    raise cherrypy.HTTPError(503, 'No MON connection')
+
+                if instance.collect_time < instance.scrape_interval:
+                    # Respond if cache isn't stale
+                    return respond()
+
+                if instance.stale_cache_strategy == instance.STALE_CACHE_RETURN:
+                    # Respond even if cache is stale
+                    instance.log.info(
+                        'Purposely responding with possibly stale Prometheus metrics cache. '
+                        'This behavior is configured in `stale_cache_strategy` configuration '
+                        'option. '
+                        'Metrics are considered stale when fetching the data takes longer than the '
+                        'configured scrape interval. '
+                        'Gathering data took {:.2f} seconds. '
+                        'Scrape interval is configured to be {:.0f} seconds.'.format(
+                            instance.collect_time,
+                            instance.scrape_interval
+                        )
+                    )
+                    return respond()
+
+                if instance.stale_cache_strategy == instance.STALE_CACHE_FAIL:
+                    # Fail if cache is stale
+                    msg = (
+                        'Prometheus metrics are stale: {:.2f} seconds too old, returning "service '
+                        'unavailable". '
+                        'This behavior is configured in `stale_cache_strategy` configuration '
+                        'option. '
+                        'Gathering data took {:.2f} seconds. '.format(
```

Comment:
Can we make these messages shorter? I don't think we want to fill the mgr log just with these messages. It should be enough to just warn about metrics being stale, we don't need to mention that the behaviour is configurable in the log, you have already mentioned it in the documentation.

Reply:
Indeed, they are very detailed. I'll shorten them.

Reply:
Would you want to keep the information about which conditions were met that led to this message, like time it took to gather the data and configured scrape interval or leave that out, too, and possibly only have that information added as debug information when debugging is enabled?

Reply:
Something like

```
Gathering data took {:.2f} seconds, metrics are stale for {:2f} seconds, returning "service unavailable".
```

and

```
Gathering data took {:.2f} seconds, metrics are stale for {:2f} seconds, returning metrics from stale cache.
```

should be fine.